examples/src/bin/gio_task.rs

```diff
@@ -0,0 +1,263 @@
+            .unwrap()
+            .propagate_value()
+            .unwrap()
+            .get::<i64>()
+            .unwrap()
+            .unwrap()
+    }
+
+    /// # Safety
+    ///
+    /// Simple getter
+    #[no_mangle]
+    pub unsafe extern "C" fn my_file_size_get_retrieved_size(this: *mut FileSize) -> i64 {
+        let simple_object = super::FileSize::from_glib_borrow(this);
+        let simple_object =
+            imp::FileSize::from_instance(&simple_object.downcast_ref::<super::FileSize>().unwrap());
+        let x = *simple_object.size.borrow();
+        x.unwrap()
+    }
+}
+
+glib::wrapper! {
+    pub struct FileSize(ObjectSubclass<imp::FileSize>);
+}
+
+impl FileSize {
+    pub fn new() -> Self {
+        glib::Object::new(&[]).expect("Failed to create FileSize")
+    }
+
+    fn retrieved_size(&self) -> i64 {
+        let simple_object = imp::FileSize::from_instance(self);
+        let x = *simple_object.size.borrow();
+        x.unwrap()
+    }
+
+    pub fn get_file_size_async<Q: FnOnce(i64, &FileSize) + 'static>(
+        &self,
+        cancellable: Option<&gio::Cancellable>,
+        callback: Q,
+    ) {
+        let closure = move |result: &gio::AsyncResult, source_object: Option<&glib::Object>| {
+            let value = result
+                .downcast_ref::<gio::Task>()
+                .unwrap()
+                .propagate_value()
+                .unwrap()
+                .get::<i64>()
+                .unwrap()
+                .unwrap();
+            let source_object = source_object.unwrap().downcast_ref::<FileSize>().unwrap();
+            callback(value, &source_object);
+        };
+
+        let task = gio::Task::new(
+            Some(&self.upcast_ref::<glib::Object>()),
+            cancellable,
+            closure,
+        );
+
+        glib::MainContext::default().spawn_local(async move {
+            let size = gio::File::new_for_path("Cargo.toml")
+                .query_info_async_future("*", gio::FileQueryInfoFlags::NONE, glib::PRIORITY_DEFAULT)
+                .await
+                .unwrap()
+                .get_size();
+
+            let source_object = task
+                .upcast_ref::<gio::AsyncResult>()
+                .get_source_object()
+                .unwrap();
+
+            let source_object =
+                imp::FileSize::from_instance(&source_object.downcast_ref::<FileSize>().unwrap());
+
+            source_object.size.replace(Some(size));
+            task.return_value(&size.to_value());
+        });
+    }
+}
+
+impl Default for FileSize {
+    fn default() -> Self {
+        FileSize::new()
+    }
+}
+
+// This function mimicks what the C code using the exported async/finish API would do. It first defines a
+// callback of type GAsyncResult that internally calls get_file_size_finish to retrieve the Task
+// result, and simply prints it out. Then it invokes the my_file_size_get_file_size_async method, passing the
+// callback to it as parameter.
+async fn run_unsafe() {
```

Comment:
You're not `await`ing anywhere, so this can as well be a non-async function. It will return immediately and then asynchronously call `c_callback` at some point.